specs/AddHostObjectToScriptWithOrigins.md

```diff
@@ -0,0 +1,275 @@
+  HRESULT add_FrameDeleted(
+      [in] ICoreWebView2FrameDeletedEventHandler *eventHandler,
+      [out] EventRegistrationToken * token);
+  /// Remove an event handler previously added with add_FrameDeleted.
+  HRESULT remove_FrameDeleted([in] EventRegistrationToken token);
+}
+
+[uuid(ae64d3bc-01ba-4769-a470-a1888de6d30b), object, pointer_default(unique)]
+interface ICoreWebView2 : IUnknown {
+  /// Raised when a new iframe is created. Use the CoreWebView2Frame.add_FrameDeleted
+  /// to listen for when this iframe goes away.
+  HRESULT add_FrameCreated(
+      [in] ICoreWebView2FrameCreatedEventHandler *eventHandler,
+      [out] EventRegistrationToken * token);
+  /// Remove an event handler previously added with add_FrameCreated.
+  HRESULT remove_FrameCreated([in] EventRegistrationToken token);
+}
+
+[uuid(6c2d169c-7912-4332-a288-0aee1626759f), object, pointer_default(unique)]
+interface ICoreWebView2FrameCreatedEventHandler: IUnknown {
+  /// Provides the result for the iframe created event
+  HRESULT Invoke([in] ICoreWebView2 * sender,
+                 [in] ICoreWebView2FrameCreatedEventArgs *
+                     args);
+}
+
+// Event args for the iframe created events.
+[uuid(5fc1cd71-cd9a-4e8f-ab5f-2f134f941a3d), object, pointer_default(unique)]
+interface ICoreWebView2FrameCreatedEventArgs : IUnknown {
+  /// The frame which was created.
+  [propget] HRESULT Frame([ out, retval ] ICoreWebView2Frame **frame);
+}
+
+[uuid(8a1860c1-528f-4286-823c-23b0439340d2), object, pointer_default(unique)]
+interface ICoreWebView2FrameDeletedEventHandler: IUnknown {
+  /// Provides the result for the iframe deleted event.
+  /// No event args exist and the `args` parameter is set to `null`.
+  HRESULT Invoke([in] ICoreWebView2Frame * sender, [in] IUnknown * args);
+}
+
+[uuid(cfd00c6a-8889-47de-84b2-3016d44dbaf4), object, pointer_default(unique)]
+interface ICoreWebView2FrameNameChangedEventHandler : IUnknown {
+  /// Provides the result for the iframe name changed event.
+  /// No event args exist and the `args` parameter is set to `null`.
+  HRESULT Invoke([in] ICoreWebView2Frame * sender, [in] IUnknown * args);
+}
+```
+## .NET and WinRT
+```c#
+namespace Microsoft.Web.WebView2.Core
+{
+    /// CoreWebView2Frame provides direct access to the iframes information and use.
+     runtimeclass CoreWebView2Frame
+    {
+        /// The name of the iframe from the iframe html tag declaring it.
+        /// Calling this method fails if it is called after the iframe is deleted.
+        String Name { get; };
+
+        /// Raised when the iframe changes its window.name property
+        event Windows.Foundation.TypedEventHandler<CoreWebView2Frame, Object> NameChanged;
+        /// The FrameDeleted event is raised when the iframe corresponding to this CoreWebView2Frame
+        /// object is removed or the document containing that iframe is destroyed
+        event Windows.Foundation.TypedEventHandler<CoreWebView2Frame, Object> FrameDeleted;
```

Comment:
Is there a way to ask a frame, "Have you been deleted"? Or does the app have to keep its own Hash`Set<Frame>` to keep track of which frames have been created and not yet deleted?

```csharp
webView2.FrameCreated += (s, e) =>
{
    activeFrames.Add(e.Frame);
    e.Frame.FrameDeleted += (s, e) =>
    {
        activeFrames.Remove(s);
    };
};

string TryGetFrameName(Frame frame)
{
    if (activeFrames.Contains(frame)) return frame.Name;
    return "";
}

void TryAddHostObjectToFrame(Frame frame, string name, object o)
{
    if (activeFrames.Contains(frame)) return frame.AddHostObject(name, o);
    return;
}
...etc for every member of Frame...

```

Reply:
See above.